In [11]:
import json
from tqdm import tqdm 
import pandas as pd
from collections import Counter
import numpy as np

In [3]:
f= 'v2_mscoco_train2014_annotations.json'

In [4]:
with open(f,'rb') as g:
    print(g)
    json_data=json.loads(g.read())

<open file 'v2_mscoco_train2014_annotations.json', mode 'rb' at 0x7fb42f8dac90>


In [12]:
json_data

[(u'info',
  {u'contributor': u'VQA Team',
   u'date_created': u'2017-04-26 17:07:13',
   u'description': u'This is v2.0 of the VQA dataset.',
   u'url': u'http://visualqa.org',
   u'version': u'2.0',
   u'year': 2017}),
 (u'data_subtype', u'train2014'),
 (u'data_type', u'mscoco'),
 (u'annotations',
  [{u'answer_type': u'other',
    u'answers': [{u'answer': u'net',
      u'answer_confidence': u'maybe',
      u'answer_id': 1},
     {u'answer': u'net', u'answer_confidence': u'yes', u'answer_id': 2},
     {u'answer': u'net', u'answer_confidence': u'yes', u'answer_id': 3},
     {u'answer': u'netting', u'answer_confidence': u'yes', u'answer_id': 4},
     {u'answer': u'net', u'answer_confidence': u'yes', u'answer_id': 5},
     {u'answer': u'net', u'answer_confidence': u'yes', u'answer_id': 6},
     {u'answer': u'mesh', u'answer_confidence': u'maybe', u'answer_id': 7},
     {u'answer': u'net', u'answer_confidence': u'yes', u'answer_id': 8},
     {u'answer': u'net', u'answer_confidence': u'yes

In [5]:
f2= 'v2_OpenEnded_mscoco_train2014_questions.json'

In [6]:
with open(f2,'r') as d:
    js= json.loads(d.read())

In [16]:
js

{u'data_subtype': u'train2014',
 u'data_type': u'mscoco',
 u'info': {u'contributor': u'VQA Team',
  u'date_created': u'2017-04-26 17:07:13',
  u'description': u'This is v2.0 of the VQA dataset.',
  u'url': u'http://visualqa.org',
  u'version': u'2.0',
  u'year': 2017},
 u'license': {u'name': u'Creative Commons Attribution 4.0 International License',
  u'url': u'http://creativecommons.org/licenses/by/4.0/'},
 u'questions': [{u'image_id': 458752,
   u'question': u'What is this photo taken looking through?',
   u'question_id': 458752000},
  {u'image_id': 458752,
   u'question': u'What position is this man playing?',
   u'question_id': 458752001},
  {u'image_id': 458752,
   u'question': u'What color is the players shirt?',
   u'question_id': 458752002},
  {u'image_id': 458752,
   u'question': u'Is this man a professional baseball player?',
   u'question_id': 458752003},
  {u'image_id': 262146,
   u'question': u'What color is the snow?',
   u'question_id': 262146000},
  {u'image_id': 262146

In [7]:
js['data_subtype']

u'train2014'

In [8]:
js.keys()

[u'info',
 u'task_type',
 u'data_type',
 u'license',
 u'data_subtype',
 u'questions']

In [9]:
js['questions'][1]

{u'image_id': 458752,
 u'question': u'What position is this man playing?',
 u'question_id': 458752001}

In [10]:
from tqdm import tqdm

In [12]:
# Insert image_id, question_id question and answer into a csv


ques_ans=[]

for i,j in tqdm(zip(js['questions'],json_data['annotations'])):
    img_id=i['image_id']
    question=i['question']
    ques_id=i['question_id']
    assert j['question_id']==ques_id , "Question ID mismatch"
    one_ans=[ans['answer'] for ans in j['answers']][0]
    ques_ans.append([img_id,ques_id,question.encode('utf-8'),one_ans.encode('utf-8')]) # encode('utf-8') because str() fails for unicode                 

100%|██████████| 443757/443757 [00:05<00:00, 86021.04it/s] 


In [65]:
len(ques_ans)

443757

In [13]:
ques_ans[10]

[393221, 393221000, 'Is the sky blue?', 'yes']

In [71]:
out_df=pd.DataFrame(ques_ans,columns=['image_id','ques_id','question','answer'])

In [72]:
out_df.to_csv('train_data.csv',index=False)

## Select only top n (n=1000) classes

In [4]:
df = pd.read_csv("../data/train_data.csv")
df.head()

image_id    ques_id                                     question   answer
0    458752  458752000    What is this photo taken looking through?      net
1    458752  458752001           What position is this man playing?  pitcher
2    458752  458752002             What color is the players shirt?   orange
3    458752  458752003  Is this man a professional baseball player?      yes
4    262146  262146000                      What color is the snow?    white

In [5]:
answers = []
for _,row in tqdm(df.iterrows()):
    answers.append(str(row.answer))

443757it [00:33, 13424.35it/s]


In [6]:
print(len(answers))
answers[:10]

443757


['net',
 'pitcher',
 'orange',
 'yes',
 'white',
 'skiing',
 'red',
 'frisbee',
 'yes',
 'frisbee']

In [12]:
len(np.unique(answers))

36255

In [66]:
freq = Counter(answers).most_common(1000)
freq

[('no', 83395),
 ('yes', 82635),
 ('2', 12289),
 ('1', 12079),
 ('white', 8139),
 ('3', 6173),
 ('red', 4879),
 ('black', 4751),
 ('blue', 4680),
 ('0', 4623),
 ('4', 4393),
 ('green', 3526),
 ('brown', 3159),
 ('yellow', 2552),
 ('5', 2310),
 ('gray', 1972),
 ('6', 1687),
 ('nothing', 1560),
 ('baseball', 1535),
 ('frisbee', 1499),
 ('tennis', 1492),
 ('orange', 1378),
 ('right', 1355),
 ('left', 1340),
 ('wood', 1294),
 ('bathroom', 1131),
 ('pizza', 1100),
 ('none', 1073),
 ('pink', 1065),
 ('kitchen', 1042),
 ('7', 1022),
 ('8', 956),
 ('cat', 888),
 ('dog', 858),
 ('skiing', 836),
 ('grass', 812),
 ('man', 777),
 ('water', 777),
 ('skateboarding', 772),
 ('silver', 753),
 ('10', 724),
 ('kite', 714),
 ('black and white', 690),
 ('skateboard', 663),
 ('surfing', 661),
 ('horse', 661),
 ('giraffe', 631),
 ('snow', 610),
 ('9', 564),
 ('surfboard', 556),
 ('wii', 556),
 ('living room', 539),
 ('tan', 537),
 ('cake', 532),
 ('elephant', 516),
 ('broccoli', 516),
 ('apple', 505),
 ('st

In [67]:
top_1000_ans=[]
sum = 0
for val,c in freq:
    sum+=c
    top_1000_ans.append(val)

print(round(sum*100/len(answers),2) ,"%")

83.14 %


There are 36255 classes in total, we can't proceed with those many class classification hence we only consider top 1000 classes as it covers about 83% of the entire data

In [68]:
new_data = []
c=0
for _,row in tqdm(df.iterrows()):
    if row.answer in top_1000_ans:
        new_data.append([row.image_id,row.ques_id,row.question,row.answer])
    else:
        c+=1
    
        


443757it [00:51, 8586.71it/s]


In [69]:
assert len(new_data)==sum, "Few entries missing check again" #to check if we got the required entries

AssertionError: Few entries missing check again

In [70]:
print(len(new_data))
sum

368900


368929

## need to check where these 29 entries are missing

In [71]:
print(len(new_data))
new_data[:5]

368900


[[458752, 458752000, 'What is this photo taken looking through?', 'net'],
 [458752, 458752001, 'What position is this man playing?', 'pitcher'],
 [458752, 458752002, 'What color is the players shirt?', 'orange'],
 [458752, 458752003, 'Is this man a professional baseball player?', 'yes'],
 [262146, 262146000, 'What color is the snow?', 'white']]

In [73]:
out_df=pd.DataFrame(new_data,columns=['image_id','ques_id','question','answer'])

In [75]:
out_df.to_csv('../data/train_data_1000.csv',index=False)

# Don't run this

In [ ]:
foo=json_data['annotations']
ques_ans=[]
for i in tqdm(js['questions']):
    img_id=i['image_id']
    question=i['question']
    ques_id=i['question_id']
    #print(img_id,ques_id,question)
    ques_ans_in=[]
    for j in foo:
        if((j['image_id']==img_id) & (j['question_id']==ques_id)):
            answers_l=[]
            for k in j['answers']:
                #print(k['answer'])
                answers_l.append(k['answer'])
            #print(answers_l)
        ques_ans_in=[img_id,ques_id,question]+answers_l
    ques_ans.append(ques_ans_in)





  0%|          | 0/443757 [00:00<?, ?it/s]



  0%|          | 1/443757 [00:00<106:52:28,  1.15it/s]



  0%|          | 2/443757 [00:01<105:43:36,  1.17it/s]



  0%|          | 3/443757 [00:02<98:55:38,  1.25it/s] 



  0%|          | 4/443757 [00:03<93:21:21,  1.32it/s]



  0%|          | 5/443757 [00:03<93:50:36,  1.31it/s]



  0%|          | 6/443757 [00:04<90:52:31,  1.36it/s]



  0%|          | 7/443757 [00:05<88:54:12,  1.39it/s]



  0%|          | 8/443757 [00:06<93:59:20,  1.31it/s]



  0%|          | 9/443757 [00:06<97:48:35,  1.26it/s]



  0%|          | 10/443757 [00:07<100:52:55,  1.22it/s]



  0%|          | 11/443757 [00:08<101:49:13,  1.21it/s]



  0%|          | 12/443757 [00:09<102:16:17,  1.21it/s]



  0%|          | 13/443757 [00:10<104:16:39,  1.18it/s]



  0%|          | 14/443757 [00:11<104:43:39,  1.18it/s]



  0%|          | 15/443757 [00:12<105:34:43,  1.17it/s]



  0%|          | 16/443757 [00:12<106:26:10,  1.16it/s]



  0%|          | 17/4

In [ ]:
ques_ans

In [ ]:
out_df=pd.DataFrame(ques_ans,columns=['image_id','ques_id','question']+['answer']*10)

In [ ]:
out_df.to_csv('out.csv')

{u'annotations': [{u'answer_type': u'other',
   u'answers': [{u'answer': u'net',
     u'answer_confidence': u'maybe',
     u'answer_id': 1},
    {u'answer': u'net', u'answer_confidence': u'yes', u'answer_id': 2},
    {u'answer': u'net', u'answer_confidence': u'yes', u'answer_id': 3},
    {u'answer': u'netting', u'answer_confidence': u'yes', u'answer_id': 4},
    {u'answer': u'net', u'answer_confidence': u'yes', u'answer_id': 5},
    {u'answer': u'net', u'answer_confidence': u'yes', u'answer_id': 6},
    {u'answer': u'mesh', u'answer_confidence': u'maybe', u'answer_id': 7},
    {u'answer': u'net', u'answer_confidence': u'yes', u'answer_id': 8},
    {u'answer': u'net', u'answer_confidence': u'yes', u'answer_id': 9},
    {u'answer': u'net', u'answer_confidence': u'yes', u'answer_id': 10}],
   u'image_id': 458752,
   u'multiple_choice_answer': u'net',
   u'question_id': 458752000,
   u'question_type': u'what is this'},
  {u'answer_type': u'other',
   u'answers': [{u'answer': u'pitcher',
  

True

458752000
458752001
458752002
458752003
